File: src/leetCode/c++/Solution271.cpp
//encode(): O(n) O(n), decode(): O(n) O(1)
class Solution271 {//lintcode659
public:
    string encode(vector<string> &strs) {
        string res = "";
        for(string str : strs)
            res += to_string(str.length()) + '$' + str;

        return res;
    }

    vector<string> decode(string &str) {
        vector<string> res;
        int len = 0;
        for(int i=0; i<str.length(); ++i){
            if(str[i] == '$'){
                string s = str.substr(i + 1, len);
                res.push_back(s);
                i += len;
                len = 0;
            }else{
                len = len * 10 + str[i] - '0';
            }
        }
        return res;
    }
};

File: java/Solution271.java
package leetCode.java;

import java.util.*;

//encode(): O(n) O(n), decode(): O(n) O(1)
class Solution271 {//lintcode659
    // Length + Delimiter: "4$lint4$code4$love3$you"
    public String encode(List<String> strs) {
        StringBuilder sb = new StringBuilder();
        for (String str : strs) {
            sb.append(str.length())
                .append("$")
                .append(str);
        }
        return sb.toString();
    }

    public List<String> decode(String str) {
        List<String> res = new ArrayList<>();
        int l = 0;
        while (l < str.length()) {
            int r = l;
            while (str.charAt(r) != '$') {
                r++;
            }

            // 長度會在 [l,r) 之間
            int len = Integer.parseInt(str.substring(l, r));
            res.add(str.substring(r + 1, r + 1 + len));
            l = r + 1 + len; 
        }

        return res;
    }
}